src/Kestrel.Transport.Sockets/Internal/SocketSender.cs

```diff
@@ -0,0 +1,108 @@
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO.Pipelines;
+using System.Net.Sockets;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
+{
+    public class SocketSender
+    {
+        private readonly Socket _socket;
+        private readonly SocketAsyncEventArgs _eventArgs = new SocketAsyncEventArgs();
+        private readonly SocketAwaitable _awaitable = new SocketAwaitable();
+
+        private List<ArraySegment<byte>> _bufferList;
+
+        public SocketSender(Socket socket)
+        {
+            _socket = socket;
+            _eventArgs.UserToken = this;
+            _eventArgs.Completed += (_, e) => ((SocketSender)e.UserToken).SendCompleted(e);
+        }
+
+        public SocketAwaitable SendAsync(ReadableBuffer buffers)
+        {
+            if (buffers.IsSingleSpan)
+            {
+                return SendAsync(buffers.First);
+            }
+
+            _eventArgs.BufferList = GetBufferList(buffers);
+
+            if (!_socket.SendAsync(_eventArgs))
+            {
+                SendCompleted(_eventArgs);
+            }
+
+            return _awaitable;
+        }
+
+        private SocketAwaitable SendAsync(Buffer<byte> buffer)
+        {
+            var segment = GetArraySegment(buffer);
+
+            _eventArgs.SetBuffer(segment.Array, segment.Offset, segment.Count);
+
+            if (!_socket.SendAsync(_eventArgs))
+            {
+                SendCompleted(_eventArgs);
+            }
+
+            return _awaitable;
+        }
+
+        private void SendCompleted(SocketAsyncEventArgs e)
+        {
+            // Clear buffer(s) to prevent the SetBuffer buffer and BufferList from both being
+            // set for the next write operation. This is unnecessary for reads since they never
+            // set BufferList.
+
+            if (e.BufferList != null)
+            {
+                e.BufferList.Clear();
+                e.BufferList = null;
+            }
+            else
+            {
+                e.SetBuffer(null, 0, 0);
+            }
+
+            _awaitable.Complete(e.BytesTransferred, e.SocketError);
+        }
+
+        private List<ArraySegment<byte>> GetBufferList(ReadableBuffer buffer)
+        {
+            Debug.Assert(!buffer.IsEmpty);
+            Debug.Assert(!buffer.IsSingleSpan);
+
+            if (_bufferList == null)
+            {
+                _bufferList = new List<ArraySegment<byte>>();
+            }
+
+            // We should always clear the list after the send
+            Debug.Assert(_bufferList.Count == 0);
+
+            foreach (var b in buffer)
+            {
+                _bufferList.Add(GetArraySegment(b));
+            }
+
+            return _bufferList;
+        }
+
+        private static ArraySegment<byte> GetArraySegment(Buffer<byte> buffer)
+        {
+            if (!buffer.TryGetArray(out var segment))
+            {
+                throw new InvalidOperationException();
```

Comment:
If this is something that we think could happen, let's at least put a message in place (even if it's a little obscure). If it's never going to happen unless we screw something up, maybe an Assert is better?

Reply:
There is an extension method for this: https://github.com/aspnet/KestrelHttpServer/blob/659fa967a1900653f7a82f02624c7c7995a3b786/src/Kestrel.Core/Internal/Http/PipelineExtensions.cs#L27